lib/std/math/gcd.zig

```diff
@@ -0,0 +1,41 @@
+// Greatest common divisor (https://mathworld.wolfram.com/GreatestCommonDivisor.html)
+const std = @import("std");
+const expectEqual = std.testing.expectEqual;
+
+/// Returns the greatest common divisor (GCD) of two integers (a, b), which are not all zero.
+/// For example, the GCD of 8 and 12 is 4, that is, gcd(8, 12) == 4.
+pub fn gcd(a: anytype, b: anytype) @TypeOf(a, b) {
+
+    // only integers are allowed and not both must be zero
+    std.debug.assert(a != 0 or b != 0);
+    std.debug.assert(a >= 0 and b >= 0);
```

Comment:
I'm thinking using `@typeInfo` it should be possible to make sure at comptime that both a and b are unsigned.

Reply:
Thank you for the suggestion. I think this is a little more elegant:
```zig
    comptime switch (@typeInfo(@TypeOf(a, b))) {
        .Int => |int| std.debug.assert(int.signedness == .unsigned),
        .ComptimeInt => {
            std.debug.assert(a >= 0);
            std.debug.assert(b >= 0);
            std.debug.assert(a != 0 or b != 0);
        },
        else => unreachable,
    };
```

Reply:
Sadly it doesn't look like that catches cases like `gcd(@as(u32, 5), -1)`.

Reply:
Ok, I hadn't actually considered that case. I'll test it again in a moment.

Reply:
But I understand what you mean, it's misleading, you don't realise the actual mistake.

Reply:
Only if you separate the type check the error is not more understandable: 
```zig
error: reached unreachable code
called from here
        .ComptimeInt => std.debug.assert(b >= 0),
```
That's why I think the "other" mistake is somehow more understandable.

Reply:
I have executed this:
```zig
@import("std").debug.print("{d}\n", .{gcd(@as(u32, 5), -1)});
```
and got this:
```zig
error: type 'u32' cannot represent integer value '-1'
    if (a == 0) return b;
```
That's how I would want to leave it.